### PR TITLE
POC: Form Flow Breadcrumbs

### PR DESCRIPTION
**Pull request by Klaas-Ritense**
Screenshot 2024-08-27 at 09 31 52

### CODE REVIEW

File: projects/valtimo/process-link/src/lib/components/form-flow/form-flow.component.ts [hunk 1]
```diff
@@ -26,43 +26,58 @@ import {
...
   @Input() formIoFormData: BehaviorSubject<any> | null = new BehaviorSubject<any>(null);
   @Input() formFlowInstanceId: string;
   @Output() formFlowComplete = new EventEmitter();
 
+  public readonly breadcrumbs$ = new BehaviorSubject<Step[]>([]);
   public readonly disabled$ = new BehaviorSubject<boolean>(false);
   public readonly formFlowStepType$ = new BehaviorSubject<FormFlowStepType | null>(null);
   public readonly FormFlowCustomComponentId$ = new BehaviorSubject<string>('');
+  public readonly currentStepIndex$ = new BehaviorSubject<number>(0);
+  public readonly enableFormFlowBreadcrumbs!: boolean;
```
Comment (mbritense) [inline projects/valtimo/process-link/src/lib/components/form-flow/form-flow.component.ts]: use observable feature toggle function so you don't have to set it in config service

File: projects/valtimo/process-link/src/lib/services/form-flow.service.ts [hunk 2]
```diff
@@ -82,4 +83,22 @@ export class FormFlowService {
       submissionData
     );
   }
+
+  getBreadcrumbs(formFlowInstanceId: string): Observable<FormFlowBreadcrumbs> {
```
Comment (mbritense) [inline projects/valtimo/process-link/src/lib/services/form-flow.service.ts]: public

File: projects/valtimo/process-link/src/lib/services/form-flow.service.ts [hunk 3]
```diff
@@ -82,4 +83,22 @@ export class FormFlowService {
...
+    return this.http.get<FormFlowBreadcrumbs>(
+      `${this.valtimoEndpointUri}v1/form-flow/instance/${formFlowInstanceId}/breadcrumbs`
+    );
+  }
+
+  navigateToStep(
```
Comment (mbritense) [inline projects/valtimo/process-link/src/lib/services/form-flow.service.ts]: public

File: projects/valtimo/process-link/src/lib/components/form-flow/form-flow.component.ts (hunk 1 above)
Comment (Klaas-Ritense) [reply projects/valtimo/process-link/src/lib/components/form-flow/form-flow.component.ts]: Changed

File: projects/valtimo/process-link/src/lib/services/form-flow.service.ts (hunk 3 above)
Comment (Klaas-Ritense) [reply projects/valtimo/process-link/src/lib/services/form-flow.service.ts]: Changed

File: projects/valtimo/process-link/src/lib/services/form-flow.service.ts (hunk 2 above)
Comment (Klaas-Ritense) [reply projects/valtimo/process-link/src/lib/services/form-flow.service.ts]: Changed

File: projects/valtimo/process-link/src/lib/components/form-flow/form-flow.component.ts [hunk 4]
```diff
@@ -26,43 +26,58 @@ import {
...
+  public readonly currentStepIndex$ = new BehaviorSubject<number>(0);
+  public readonly enableFormFlowBreadcrumbs$: Observable<boolean>;
+
+  private readonly _subscriptions = new Subscription();
 
   formDefinition: FormioForm;
   formioOptions: ValtimoFormioOptions;
 
-  private formFlowStepInstanceId: string;
+  formFlowStepInstanceId: string;
```
Comment (teodor-ritense) [inline projects/valtimo/process-link/src/lib/components/form-flow/form-flow.component.ts]: Why was this private removed?

File: projects/valtimo/process-link/src/lib/components/form-flow/form-flow.component.ts [hunk 5]
```diff
@@ -26,43 +26,58 @@ import {
...
 
   constructor(
     private readonly formFlowService: FormFlowService,
-    private readonly modalService: ValtimoModalService
+    private readonly modalService: ValtimoModalService,
+    private readonly translateService: TranslateService,
+    configService: ConfigService
```
Comment (teodor-ritense) [inline projects/valtimo/process-link/src/lib/components/form-flow/form-flow.component.ts]:
```suggestion
    private readonly configService: ConfigService
```

File: projects/valtimo/process-link/src/lib/components/form-flow/form-flow.component.ts [hunk 6]
```diff
@@ -106,6 +123,56 @@ export class FormFlowComponent implements OnInit {
...
+        this.formFlowStepInstanceId,
+        event.step.stepInstanceId,
+        submissionData
+      )
+      .subscribe(
+        (result: FormFlowInstance) => this.handleFormFlowStep(result),
+        errors => {
+          this.form?.showErrors(errors);
+          this.enable();
+        }
```
Comment (teodor-ritense) [inline projects/valtimo/process-link/src/lib/components/form-flow/form-flow.component.ts]:
```suggestion
        {
        next: (result: FormFlowInstance) => this.handleFormFlowStep(result),
        error: errors => {
          this.form?.showErrors(errors);
          this.enable();
        }
        }
```

File: projects/valtimo/process-link/src/lib/components/form-flow/form-flow.component.ts (hunk 4 above)
Comment (Klaas-Ritense) [reply projects/valtimo/process-link/src/lib/components/form-flow/form-flow.component.ts]: No reason at al :p

File: projects/valtimo/process-link/src/lib/components/form-flow/form-flow.component.ts [hunk 7]
```diff
@@ -26,21 +26,30 @@ import {
...
   @Input() formIoFormData: BehaviorSubject<any> | null = new BehaviorSubject<any>(null);
   @Input() formFlowInstanceId: string;
   @Output() formFlowComplete = new EventEmitter();
 
+  public readonly breadcrumbs$ = new BehaviorSubject<Step[]>([]);
   public readonly disabled$ = new BehaviorSubject<boolean>(false);
   public readonly formFlowStepType$ = new BehaviorSubject<FormFlowStepType | null>(null);
   public readonly FormFlowCustomComponentId$ = new BehaviorSubject<string>('');
+  public readonly currentStepIndex$ = new BehaviorSubject<number>(0);
+  public readonly enableFormFlowBreadcrumbs$: Observable<boolean>;
```
Comment (teodor-ritense) [inline projects/valtimo/process-link/src/lib/components/form-flow/form-flow.component.ts]:
```suggestion
  public readonly enableFormFlowBreadcrumbs$ = this.configService.getFeatureToggleObservable('enableFormFlowBreadcrumbs');;
```

File: projects/valtimo/process-link/src/lib/components/form-flow/form-flow.component.ts [hunk 8]
```diff
@@ -49,20 +58,26 @@ export class FormFlowComponent implements OnInit {
...
   constructor(
     private readonly formFlowService: FormFlowService,
-    private readonly modalService: ValtimoModalService
+    private readonly modalService: ValtimoModalService,
+    private readonly translateService: TranslateService,
+    private readonly configService: ConfigService
   ) {
     this.formioOptions = new FormioOptionsImpl();
     this.formioOptions.disableAlerts = true;
+    this.enableFormFlowBreadcrumbs$ = configService.getFeatureToggleObservable('enableFormFlowBreadcrumbs');
```
Comment (teodor-ritense) [inline projects/valtimo/process-link/src/lib/components/form-flow/form-flow.component.ts]: This can be removed with the above change

**Review by teodor-ritense — APPROVED** [review]

**Pull request merged**